internal/cli/demo/k3d.go

```diff
@@ -0,0 +1,165 @@
+	}
+
+	return &K3dClient{
+		cli:      cli,
+		k3dImage: ref,
+	}
+}
+
+func (k *K3dClient) ListClusters(ctx context.Context) ([]string, error) {
+	cmd := []string{"cluster", "list", "-ojson"}
+	var clusterListJson bytes.Buffer
+	stderr := logger.Get(ctx).Writer(logger.WarnLvl)
+	if err := k.command(ctx, cmd, &clusterListJson, stderr, true); err != nil {
+		return nil, err
+	}
+
+	var clusters []cluster
+	if err := json.Unmarshal(clusterListJson.Bytes(), &clusters); err != nil {
+		return nil, fmt.Errorf("invalid JSON output from cluster list: %v", err)
+	}
+
+	clusterNames := make([]string, len(clusters))
+	for i := range clusters {
+		clusterNames[i] = clusters[i].Name
+	}
+	return clusterNames, nil
+}
+
+func (k *K3dClient) CreateCluster(ctx context.Context, clusterName string) error {
+	cmd := []string{
+		"cluster",
+		"create", clusterName,
+		"--registry-create",
+		"--kubeconfig-switch-context",
+		"--kubeconfig-update-default",
+		"--no-hostip",
+		"--no-image-volume",
+		"--no-lb",
+		"--label", fmt.Sprintf("%s@%s", docker.BuiltByTiltLabelStr, "server[0]"),
+	}
+	stdout := logger.Get(ctx).Writer(logger.DebugLvl)
+	stderr := logger.Get(ctx).Writer(logger.WarnLvl)
+	if err := k.command(ctx, cmd, stdout, stderr, true); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (k *K3dClient) DeleteCluster(ctx context.Context, clusterName string, wait bool) error {
+	cmd := []string{
+		"cluster",
+		"delete", clusterName,
+	}
+	var stdout, stderr io.Writer
+	if wait {
+		log := logger.Get(ctx)
+		stdout = log.Writer(logger.DebugLvl)
+		stderr = logger.NewFuncLogger(log.SupportsColor(), log.Level(),
+			func(level logger.Level, fields logger.Fields, b []byte) error {
+				// there's no kubeconfig in the container so k3d will emit confusing warnings
+				// note: no kubeconfig cleanup is necessary since k3d's execution is isolated
+				// 	via docker, so is never touching the host filesystem, but it's a weird
+				// 	use case so k3d doesn't have a flag to disable kubeconfig cleanup on delete
+				if bytes.Contains(b, []byte("Failed to remove cluster details")) ||
+					bytes.Contains(b, []byte("no such file or directory")) {
+					return nil
+				}
+				log.Write(logger.WarnLvl, b)
+				return nil
+			}).Writer(logger.WarnLvl)
+	}
+	if err := k.command(ctx, cmd, stdout, stderr, wait); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (k *K3dClient) GenerateKubeconfig(ctx context.Context, clusterName string) ([]byte, error) {
+	var kubeconfigBuf bytes.Buffer
+	stderr := logger.Get(ctx).Writer(logger.WarnLvl)
+	err := k.command(ctx, []string{"kubeconfig", "get", clusterName}, &kubeconfigBuf, stderr, true)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get kubeconfig: %v", err)
+	}
+	return kubeconfigBuf.Bytes(), nil
+}
+
+func (k *K3dClient) command(ctx context.Context, cmd []string, stdout io.Writer, stderr io.Writer, wait bool) error {
+	k.ensurePulled.Do(func() {
+		ref, err := k.cli.ImagePull(ctx, k.k3dImage)
```

Comment:
is there a reason why you chose to pull the image up-front rather than Pull as part of the Run?

Reply:
Just to speed things up a bit because we run multiple k3d commands, each one in an ephemeral container, so it's redundant after the first time. This isn't _hugely_ critical at startup, but on shutdown, we have < 2 sec to fire off the Docker API request before the signal handler forcibly terminates the process 😬

(For reference, it takes >500ms to `ImagePull()` on something I've already got pulled, which I'm guessing is largely dependent on local I/O to find the image digest and then the network request to registry API to compare.)